### PR TITLE
[SYCL][HIP] Add coarse-grained memory advice for HIP on AMD

### PR DESCRIPTION
**Pull request by GeorgeWeb**
Enables and tests coarse grained memory access via the memadvise implementation for HIP platforms on AMD hardware.

See related UR changes for the adapter implementation: https://github.com/oneapi-src/unified-runtime/pull/1249

### CODE REVIEW

File: sycl/test-e2e/USM/memory_coherency_hip.cpp
Comment (steffenlarsen) [inline sycl/test-e2e/USM/memory_coherency_hip.cpp:126]: Nit; Might as well just use [`sleep_for`.](https://en.cppreference.com/w/cpp/thread/sleep_for)

**Review by steffenlarsen — APPROVED** [review]
Overall I think it looks solid. Converting to draft until UR changes have been merged though.

File: sycl/test-e2e/USM/memory_coherency_hip.cpp
Comment (GeorgeWeb) [reply sycl/test-e2e/USM/memory_coherency_hip.cpp:126]: Cheers, yeah might as well. Will get back to updating it after UR reviews are done. Thanks!

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:6]:
```suggestion
```

**Review by kbenzie — APPROVED** [review]
UR LGTM once the conflicts are resolved and *DO-NOT-MERGE* line is deleted.

**Comment by GeorgeWeb** [comment]
@jchlanda May I have a final look at this one, please? Thanks!

**Comment by GeorgeWeb** [comment]
@intel/llvm-gatekeepers This one looks ready for merging now. 🙏 The UR changes for it got in on Friday.

File: sycl/plugins/unified_runtime/pi2ur.hpp
Comment (jchlanda) [inline sycl/plugins/unified_runtime/pi2ur.hpp:3378]: I'm not sure how `UrAdvice` is used later on (so feel free to disregard this comment), but I'm wondering if it makes sense for "toggle" options (such as switch on/off coarse grain mem advice) to also unset the opposite value. Since the values of toggles are not exclusive (10th and 11th bit flipped on for mem advice) you could end up in a situation when both are set and it'd be down to implementation to decide what to do?

As in, how would the implementation react to this (slightly forced bit of code):
```cpp
q.mem_advise(ptr, sizeof(int), int{PI_MEM_ADVICE_HIP_SET_COARSE_GRAINED});
q.mem_advise(ptr, sizeof(int), int{PI_MEM_ADVICE_HIP_UNSET_COARSE_GRAINED});
```
My worry is that somewhere there is a snippet like:
```cpp
if (UrAdvice & PI_MEM_ADVICE_HIP_SET_COARSE_GRAINED)
  do blah
```
and, if I understand it correctly `do blah` would now get triggered, no?

**Review by jchlanda — APPROVED** [review]

File: sycl/test-e2e/USM/memory_coherency_hip.cpp
Comment (steffenlarsen) [reply sycl/test-e2e/USM/memory_coherency_hip.cpp:126]: Quick reminder on this one, now that we are close to merge.

File: sycl/plugins/unified_runtime/pi2ur.hpp
Comment (GeorgeWeb) [reply sycl/plugins/unified_runtime/pi2ur.hpp:3378]: Thanks for raising this. It's been used like this for a while, before these changes. I can see the concern but in the adapters each URAdvice set flag is mapped to a underlying value. Each value is then processed in a loop, preserving the natural set/unset order.

Any other use-case of URAdvice in our adapters that is outside of the usm madvise API might suffer from the flakiness of this design. If the need occurs in future, we may follow-up to resolve this.

File: sycl/test-e2e/USM/memory_coherency_hip.cpp
Comment (GeorgeWeb) [reply sycl/test-e2e/USM/memory_coherency_hip.cpp:125]: Thanks. Just pushed the commit.

**Comment by aarongreig** [comment]
@intel/llvm-gatekeepers CI is passing now, please merge when you can

**Comment by aelovikov-intel** [comment]
Post-commit failures on Arc GPU:
```
Failed Tests (1):
  SYCL :: Regression/device_pci_address_bdf_format.cpp
```
```
FAIL: SYCL :: Regression/device_pci_address_bdf_format.cpp (1557 of 1869)
******************** TEST 'SYCL :: Regression/device_pci_address_bdf_format.cpp' FAILED ********************
Exit Code: -6

Command Output (stdout):
--
# RUN: at line 2
/__w/llvm/llvm/toolchain/bin//clang++   -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/Regression/device_pci_address_bdf_format.cpp -o /__w/llvm/llvm/build-e2e/Regression/Output/device_pci_address_bdf_format.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/Regression/device_pci_address_bdf_format.cpp -o /__w/llvm/llvm/build-e2e/Regression/Output/device_pci_address_bdf_format.cpp.tmp.out
# .---command stderr------------
# | /__w/llvm/llvm/llvm/sycl/test-e2e/Regression/device_pci_address_bdf_format.cpp:43:38: warning: 'ext_intel_pci_address' is deprecated: use ext::intel::info::device::pci_address instead [-Wdeprecated-declarations]
# |    43 |           dev.get_info<info::device::ext_intel_pci_address>();
# |       |                                      ^
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/info/device_traits_deprecated.def:23:1: note: 'ext_intel_pci_address' has been explicitly marked deprecated here
# |    23 | __SYCL_PARAM_TRAITS_DEPRECATED(ext_intel_pci_address,"use ext::intel::info::device::pci_address instead")
# |       | ^
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/info/info_desc.hpp:111:10: note: expanded from macro '__SYCL_PARAM_TRAITS_DEPRECATED'
# |   111 |   struct __SYCL2020_DEPRECATED(Message) Desc;
# |       |          ^
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/defines_elementary.hpp:53:40: note: expanded from macro '__SYCL2020_DEPRECATED'
# |    53 | #define __SYCL2020_DEPRECATED(message) __SYCL_DEPRECATED(message)
# |       |                                        ^
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/defines_elementary.hpp:44:38: note: expanded from macro '__SYCL_DEPRECATED'
# |    44 | #define __SYCL_DEPRECATED(message) [[deprecated(message)]]
# |       |                                      ^
# | 1 warning generated.
# | /__w/llvm/llvm/llvm/sycl/test-e2e/Regression/device_pci_address_bdf_format.cpp:43:38: warning: 'ext_intel_pci_address' is deprecated: use ext::intel::info::device::pci_address instead [-Wdeprecated-declarations]
# |    43 |           dev.get_info<info::device::ext_intel_pci_address>();
# |       |                                      ^
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/info/device_traits_deprecated.def:23:1: note: 'ext_intel_pci_address' has been explicitly marked deprecated here
# |    23 | __SYCL_PARAM_TRAITS_DEPRECATED(ext_intel_pci_address,"use ext::intel::info::device::pci_address instead")
# |       | ^
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/info/info_desc.hpp:111:10: note: expanded from macro '__SYCL_PARAM_TRAITS_DEPRECATED'
# |   111 |   struct __SYCL2020_DEPRECATED(Message) Desc;
# |       |          ^
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/defines_elementary.hpp:53:40: note: expanded from macro '__SYCL2020_DEPRECATED'
# |    53 | #define __SYCL2020_DEPRECATED(message) __SYCL_DEPRECATED(message)
# |       |                                        ^
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/defines_elementary.hpp:44:38: note: expanded from macro '__SYCL_DEPRECATED'
# |    44 | #define __SYCL_DEPRECATED(message) [[deprecated(message)]]
# |       |                                      ^
# | 1 warning generated.
# `-----------------------------
# RUN: at line 3
env ONEAPI_DEVICE_SELECTOR=level_zero:gpu  /__w/llvm/llvm/build-e2e/Regression/Output/device_pci_address_bdf_format.cpp.tmp.out
# executed command: env ONEAPI_DEVICE_SELECTOR=level_zero:gpu /__w/llvm/llvm/build-e2e/Regression/Output/device_pci_address_bdf_format.cpp.tmp.out
# .---command stdout------------
# | PCI address = 0000:03:00.0
# | PCI address = 0000:00:02.0
# | Passed!
# | Abort was called at 253 line in file:
# | ../../neo/level_zero/core/source/builtin/builtin_functions_lib_impl.cpp
# `-----------------------------
# .---command stderr------------
# | pure virtual method called
# | terminate called without an active exception
# `-----------------------------
# error: command failed with exit status: -6
```

**Comment by GeorgeWeb** [comment]
> Post-commit failures on Arc GPU:
> 
> ```
> 
> Failed Tests (1):
> 
>   SYCL :: Regression/device_pci_address_bdf_format.cpp
> 
> ```
> 
> ```
> 
> FAIL: SYCL :: Regression/device_pci_address_bdf_format.cpp (1557 of 1869)
> 
> ******************** TEST 'SYCL :: Regression/device_pci_address_bdf_format.cpp' FAILED ********************
> 
> Exit Code: -6
> 
> 
> 
> Command Output (stdout):
> 
> --
> 
> # RUN: at line 2
> 
> /__w/llvm/llvm/toolchain/bin//clang++   -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/Regression/device_pci_address_bdf_format.cpp -o /__w/llvm/llvm/build-e2e/Regression/Output/device_pci_address_bdf_format.cpp.tmp.out
> 
> # executed command: /__w/llvm/llvm/toolchain/bin//clang++ -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/Regression/device_pci_address_bdf_format.cpp -o /__w/llvm/llvm/build-e2e/Regression/Output/device_pci_address_bdf_format.cpp.tmp.out
> 
> # .---command stderr------------
> 
> # | /__w/llvm/llvm/llvm/sycl/test-e2e/Regression/device_pci_address_bdf_format.cpp:43:38: warning: 'ext_intel_pci_address' is deprecated: use ext::intel::info::device::pci_address instead [-Wdeprecated-declarations]
> 
> # |    43 |           dev.get_info<info::device::ext_intel_pci_address>();
> 
> # |       |                                      ^
> 
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/info/device_traits_deprecated.def:23:1: note: 'ext_intel_pci_address' has been explicitly marked deprecated here
> 
> # |    23 | __SYCL_PARAM_TRAITS_DEPRECATED(ext_intel_pci_address,"use ext::intel::info::device::pci_address instead")
> 
> # |       | ^
> 
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/info/info_desc.hpp:111:10: note: expanded from macro '__SYCL_PARAM_TRAITS_DEPRECATED'
> 
> # |   111 |   struct __SYCL2020_DEPRECATED(Message) Desc;
> 
> # |       |          ^
> 
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/defines_elementary.hpp:53:40: note: expanded from macro '__SYCL2020_DEPRECATED'
> 
> # |    53 | #define __SYCL2020_DEPRECATED(message) __SYCL_DEPRECATED(message)
> 
> # |       |                                        ^
> 
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/defines_elementary.hpp:44:38: note: expanded from macro '__SYCL_DEPRECATED'
> 
> # |    44 | #define __SYCL_DEPRECATED(message) [[deprecated(message)]]
> 
> # |       |                                      ^
> 
> # | 1 warning generated.
> 
> # | /__w/llvm/llvm/llvm/sycl/test-e2e/Regression/device_pci_address_bdf_format.cpp:43:38: warning: 'ext_intel_pci_address' is deprecated: use ext::intel::info::device::pci_address instead [-Wdeprecated-declarations]
> 
> # |    43 |           dev.get_info<info::device::ext_intel_pci_address>();
> 
> # |       |                                      ^
> 
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/info/device_traits_deprecated.def:23:1: note: 'ext_intel_pci_address' has been explicitly marked deprecated here
> 
> # |    23 | __SYCL_PARAM_TRAITS_DEPRECATED(ext_intel_pci_address,"use ext::intel::info::device::pci_address instead")
> 
> # |       | ^
> 
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/info/info_desc.hpp:111:10: note: expanded from macro '__SYCL_PARAM_TRAITS_DEPRECATED'
> 
> # |   111 |   struct __SYCL2020_DEPRECATED(Message) Desc;
> 
> # |       |          ^
> 
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/defines_elementary.hpp:53:40: note: expanded from macro '__SYCL2020_DEPRECATED'
> 
> # |    53 | #define __SYCL2020_DEPRECATED(message) __SYCL_DEPRECATED(message)
> 
> # |       |                                        ^
> 
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/defines_elementary.hpp:44:38: note: expanded from macro '__SYCL_DEPRECATED'
> 
> # |    44 | #define __SYCL_DEPRECATED(message) [[deprecated(message)]]
> 
> # |       |                                      ^
> 
> # | 1 warning generated.
> 
> # `-----------------------------
> 
> # RUN: at line 3
> 
> env ONEAPI_DEVICE_SELECTOR=level_zero:gpu  /__w/llvm/llvm/build-e2e/Regression/Output/device_pci_address_bdf_format.cpp.tmp.out
> 
> # executed command: env ONEAPI_DEVICE_SELECTOR=level_zero:gpu /__w/llvm/llvm/build-e2e/Regression/Output/device_pci_address_bdf_format.cpp.tmp.out
> 
> # .---command stdout------------
> 
> # | PCI address = 0000:03:00.0
> 
> # | PCI address = 0000:00:02.0
> 
> # | Passed!
> 
> # | Abort was called at 253 line in file:
> 
> # | ../../neo/level_zero/core/source/builtin/builtin_functions_lib_impl.cpp
> 
> # `-----------------------------
> 
> # .---command stderr------------
> 
> # | pure virtual method called
> 
> # | terminate called without an active exception
> 
> # `-----------------------------
> 
> # error: command failed with exit status: -6
> 
> ```

Unsure how PI HIP changes relate to it, but I am assuming an effect of maybe the updated UR commit tag. I can't take immediate look, will have to be later or tomorrow if I am unable to by end of today.

**Pull request merged**
